servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,257 @@
+        requireNonNull(src);
+
+        if (offset < 0) {
+            throw new IllegalArgumentException("Invalid offset: " + offset + " (expected >= 0)");
+        }
+
+        if (length <= 0) {
+            throw new IllegalArgumentException("Invalid length: " + length + " (expected > 0)");
+        }
+
+        final ByteToMessageDecoder decoder = decoderFn.get();
+        final EmbeddedChannel channel = new EmbeddedChannel(decoder);
+
+        try {
+            ByteBuf origin = toByteBufRetained(src, offset, length);
+            channel.writeInbound(origin);
+            return newBufferFromRetained(channel.readInbound());
+        } catch (Exception e) {
+            LOGGER.error("Error while decoding with {}", name(), e);
+            throwException(e);
+            return EMPTY_BUFFER;
+        } finally {
+            safeClose(channel);
+        }
+    }
+
+    @Override
+    public Publisher<Buffer> decode(final Publisher<Buffer> from, final BufferAllocator __) {
+        requireNonNull(from);
+        return from.liftSync(subscriber -> new PublisherSource.Subscriber<Buffer>() {
+
+            private final ByteToMessageDecoder decoder = decoderFn.get();
+            private final EmbeddedChannel channel = new EmbeddedChannel(decoder);
+
+            @Nullable
+            PublisherSource.Subscription subscription;
+
+            @Override
+            public void onSubscribe(final PublisherSource.Subscription subscription) {
+                this.subscription = subscription;
+                subscriber.onSubscribe(subscription);
+            }
+
+            @Override
+            public void onNext(@Nullable final Buffer src) {
+                assert subscription != null;
+
+                if (src == null) {
+                    return;
+                }
+
+                // onNext will produce AT-MOST N items (as received)
+                if (!channel.isOpen()) {
+                    throw new IllegalStateException("Stream encoder previously closed but more input arrived ");
+                }
+
+                channel.writeInbound(toByteBufRetained(src));
+                ByteBuf part = channel.readInbound();
+
+                if (part != null && part.readableBytes() > 0) {
+                    subscriber.onNext(newBufferFromRetained(part));
+                } else {
+                    // Not enough data to decompress, ask for more
+                    subscription.request(1);
+                }
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                safeClose(channel);
+                subscriber.onError(t);
+            }
+
+            @Override
+            public void onComplete() {
+                safeClose(channel);
+                subscriber.onComplete();
+            }
+        });
+    }
+
+    private static Buffer newBufferFromRetained(final ByteBuf buf) {
+        return newBufferFrom(buf.readRetainedSlice(buf.readableBytes()));
```

Comment:
If you configure `EmbeddedChannel` to use ST allocator, you don't need to retain `ByteBuf` it returns, just use `newBufferFrom`.